File: src/014/p014.ts
const getCollatzSequenceLength = (
  n: number,
  chainLengths: { [key: number]: number }
): number => {
  if (chainLengths[n] === undefined) {
    const next = n % 2 === 0 ? n / 2 : 3 * n + 1;
    chainLengths[n] = getCollatzSequenceLength(next, chainLengths) + 1;
  }
  return chainLengths[n];
};

export const solve = (): string => {
  const chainLengths: { [key: number]: number } = {};
  chainLengths[1] = 1;

  let maxChainLength = 0;
  let maxChainLengthStart = 0;
  for (let i = 1; i < 1000000; i++) {
    const len = getCollatzSequenceLength(i, chainLengths);
    if (len > maxChainLength) {
      maxChainLength = len;
      maxChainLengthStart = i;
    }
  }
  return maxChainLengthStart.toString();
};

File: src/014/p014.py
def solve() -> int:
    chain_lengths = [0] * 1000000
    chain_lengths[1] = 1

    max_chain_length = 0
    max_chain_length_start = 0

    for start in range(500000, 1000000):
        chain_length = 0
        num = start
        while True:
            if num % 2 == 0:
                num = num // 2
                if num < start:
                    chain_lengths[start] = chain_length + chain_lengths[num]
                    break
            else:
                num = 3 * num + 1
            chain_length += 1

        if chain_lengths[start] > max_chain_length:
            max_chain_length = chain_lengths[start]
            max_chain_length_start = start
    return max_chain_length_start
